src/rsz/src/SpefWriter.cc

```diff
@@ -0,0 +1,246 @@
+  if (it == spef_streams_.end()) {
+    return;
+  }
+  std::ostream& stream = *it->second;
+  stream << "*SPEF \"ieee 1481-1999\"" << '\n';
+  stream << "*DESIGN \"" << network_->block()->getName() << "\"" << '\n';
+  stream << "*DATE \"11:11:11 Fri 11 11, 1111\"" << '\n';
+  stream << "*VENDOR \"The OpenROAD Project\"" << '\n';
+  stream << "*PROGRAM \"OpenROAD\"" << '\n';
+  stream << "*VERSION \"1.0\"" << '\n';
+  stream << "*DESIGN_FLOW \"NAME_SCOPE LOCAL\" \"PIN_CAP NONE\"" << '\n';
+  stream << "*DIVIDER /" << '\n';
+  stream << "*DELIMITER :" << '\n';
+  stream << "*BUS_DELIMITER []" << '\n';
+
+  auto units = network_->units();
+  std::string time_unit = std::string(units->timeUnit()->suffix());
+  std::string cap_unit = std::string(units->capacitanceUnit()->suffix());
+  std::string res_unit = std::string(units->resistanceUnit()->suffix());
+  std::transform(
+      time_unit.begin(), time_unit.end(), time_unit.begin(), ::toupper);
+  std::transform(cap_unit.begin(), cap_unit.end(), cap_unit.begin(), ::toupper);
+  std::transform(res_unit.begin(), res_unit.end(), res_unit.begin(), ::toupper);
+
+  stream << "*T_UNIT 1 " << time_unit << '\n';
+  stream << "*C_UNIT 1 " << cap_unit << '\n';
+  stream << "*R_UNIT 1 " << res_unit << '\n';
+  stream << '\n';
+}
+
+void SpefWriter::writeSpefPorts(Corner* corner)
+{
+  auto it = spef_streams_.find(corner);
+  if (it == spef_streams_.end()) {
+    return;
+  }
+  std::ostream& stream = *it->second;
+
+  std::unique_ptr<sta::InstancePinIterator> pin_iter(
+      network_->pinIterator(network_->topInstance()));
+
+  stream << "*PORTS" << '\n';
+  while (pin_iter->hasNext()) {
+    sta::Pin* pin = pin_iter->next();
+    odb::dbITerm* iterm = nullptr;
+    odb::dbBTerm* bterm = nullptr;
+    odb::dbModITerm* moditerm = nullptr;
+    odb::dbModBTerm* modbterm = nullptr;
+    network_->staToDb(pin, iterm, bterm, moditerm, modbterm);
+
+    if (iterm != nullptr) {
+      stream << iterm->getName() << " ";
+      if (iterm->getIoType() == odb::dbIoType::INPUT) {
+        stream << "I";
+      } else if (iterm->getIoType() == odb::dbIoType::OUTPUT) {
+        stream << "O";
+      } else {
+        stream << "B";
+      }
+      stream << '\n';
+    } else if (bterm != nullptr) {
+      stream << bterm->getName() << " ";
+      if (bterm->getIoType() == odb::dbIoType::INPUT) {
+        stream << "I";
+      } else if (bterm->getIoType() == odb::dbIoType::OUTPUT) {
+        stream << "O";
+      } else {
+        stream << "B";
+      }
+      stream << '\n';
+    } else {
+      logger_->error(
+          RSZ,
+          8,
+          "Got a modTerm instead of iTerm or bTerm while writing SPEF ports.");
+    }
+  }
+  stream << '\n';
+}
+
+void SpefWriter::writeSpefNet(Corner* corner,
+                              const Net* net,
+                              Parasitic* parasitic)
+{
+  auto it = spef_streams_.find(corner);
+  if (it == spef_streams_.end()) {
+    return;
```

Comment:
This seems like an error.

Reply:
This is here for the case where the caller wants to write only a subset of the available corners using the rsz::estimateWireParastics or grt::estimateRC functions.

Reply:
This seems like the opposite - they are writing a corner they didn't declare any ostream for.  A corner in a subset would be found.